### ISSUE TITLE
Add instructions for custom cell typing references

### ISSUE THREAD

**Issue by sjspielman** [AlexsLemonade/scpca-nf#567]
We'll need to either expand `external-instructions.md`, or write a separate docs file, to document how to create and format your own cell type reference files for `SingleR` and `CellAssign`. This should include a discussion of how to run the workflow (e.g., need to update reference paths) with custom references.

**Comment by sjspielman** [comment]
We should more carefully outline the scope of this issue to determine if/when we should pursue this, as well as how long it will realistically take.

**Comment by sjspielman** [comment]
I've sketched out an outline for what these instructions might entail. I do think it probably rises to the level of its own file?

Noting we already provide information on how to all the references we have pre-built: https://github.com/AlexsLemonade/scpca-nf/blob/main/external-instructions.md#choosing-reference-datasets, so this file would focus only on making your own reference "from scratch."

---

### Building custom references with our `build-celltype-ref.nf` workflow

Nope, don't do this, build your own.


### Building and using custom references

#### DIY `SingleR`
- You'll need to set this parameter value to directory with your stored model: `params.singler_models_dir`
- Create a reference model (SCE or SE object) with file name `<your singler reference name>_model.rds`
- Build your model as follows and then export to `.rds` file:
  ```
  singler_model <- SingleR::trainSingleR(
    ref = <your reference se or sce>, # columns are samples/cells and rows are ensembl ids
    labels = <vector of labels for each sample or cell>,
    restrict = <vector of gene ids that are in the mapping reference>
  )
  ```
  - For the `labels` argument:
	  - If you are using a `celldex` dataset for this:
		  - If you are using `label.ont`, then the workflow will expect more things, stay tuned.
		- If you are not using `label.ont`, then you will need to override the parameter with a name that represents your label: `params.singler_label_name = "your label"`
  - For the `restrict` argument, we can point them to our tx2gene file, if they are using our human reference. 

- If you used a `celldex` SE with `label.ont`, you have two choices:
	- add a data frame to the model as `singler_model$cell_ontology_df`  with two columns, `ontology_id` and `ontology_cell_names`. You may find `ontoProc::getOnto("cellonto")` useful hint hint.
	- OR override `params.singler_label_name = "don't use label.ont here"`
- You will also need to add `singler_model$reference_label <- "the label you are setting"`. Honestly this is probably optional, the only consequence is that `metadata(sce)$singler_reference_label` will end up as `NULL`.


#### DIY `CellAssign`
- You'll need to set this parameter value to directory with your stored reference: `params.cellassign_ref_dir`
- Create a reference TSV file with file name `<your cellassign reference name>.tsv`. It should have these columns:
  - `ensembl_id`
  -  ...column for each cell type...
    - each of these has a value 0 or 1 for if it's a marker gene or not.
  - Not required, but in our workflow we include an `"other"` column (name it what you want) with all 0. `CellAssign` will end up assigning cells to this cell type when none of the other cell types fit. It's your "could not assign this cell" column.


#### Make your metadata file

You'll need to create a TSV file and send it into `params.project_celltype_metafile`. You'll need the following columns:
- `scpca_project_id`: Each sample/library in this project gets the same celltype references
- `singler_ref_name`: Name it something informative for you. This should match `<your singler reference name>` in the filename.
- `singler_ref_file`: Name of your SingleR model file, sans path.
- `cellassign_ref_name`: Name it something informative for you. This should match `<your cellassign reference name>` in the filename.
- `cellassign_ref_file`: Name of your CellAssign reference file, sans path.


#### Run the workflow

```
nextflow run AlexsLemonade/scpca-nf \
  ... your usual flags.... \
  --singler_label_name <optionally, the label name> \ 
  --singler_models_dir <path/to/singler/model/file> \
  --cellassign_reference/dir <path/to/reference/file> \
  --project_celltype_metafile <path/to/celltype metafile> \ 
  --perform_celltyping
```

We use caching based around the model/reference file names. If results are already in the checkpoint (show file tree) for that file name, then it will be skipped. To re-do celltyping with the same file name, toss in `--repeat_celltyping`.

**Comment by sjspielman** [comment]
Closed by #692
